meson.build

```diff
@@ -7,48 +7,71 @@ project(
     'warning_level=3',
     'cpp_std=c++14',
     'b_ndebug=if-release',
-    'default_library=both',
   ],
   license : 'MIT',
   meson_version : '>= 0.52.0',
 )
 
-edlib_lib = library('edlib',
-  sources : files(['edlib/src/edlib.cpp']),
-  include_directories : ['edlib/include'],
+project_version_major = meson.project_version().split('.')[0]
+
+###### Libraries ######
+
+edlib_lib_sources = files(['edlib/src/edlib.cpp'])
+edlib_lib_include_dirs = include_directories('edlib/include')
+
+edlib_lib_static = static_library('edlib',
+  sources : edlib_lib_sources,
+  include_directories : edlib_lib_include_dirs,
   dependencies : [],
   install : true,
 )
+edlib_dep_static = declare_dependency(
+  include_directories : edlib_lib_include_dirs,
+  link_with : edlib_lib_static
+)
```

Comment:
the problem here: it perpetuates the cmake anti-patterns (i.e. always building two libraries). Here's my simple suggestion:
1. go back to `edlib_lib`
2. but do not add back `'default_library=both'`
3. either specify `'default_library=static'` or (better IMO) tell meson to use static libraries by default in Windows.

The fundamental issue here is that Windows doesn't have a good `RPATH` mechanism, which means that binaries linking to shared libraries on Windows for testing is always a pain.

Reply:
For the maintainer/you, `'default_library=static'` is easier to manage, for distros, `'default_library=shared'`/omitting it entirely is easier, even though most distros can just change the option themselves.

Reply:
Ok, if that is best practice let's do it like that!

Ok so I could set it to static by default in WIndows and shared by default if not Windows. How would I do that, do you have any advice?
Also, if I use just library(), how do I specify additional c_flags when the library is shared? There are couple of places where I need to specify them through the meson.build. Would you just do it conditionally? What would you check in the condition, 'default_library'? But what if it is set to 'both'?

RPATH and Windows -> It actually worked on appveyor! I linked runTests with dynamic library and it works fine. But maybe that is exception or I completely misunderstood you?

From my side as edlib maintainer, I see following goals:
1. We want to ensure both static and dynamic edlib libraries are correctly built on Linux/OSX/Windows. I want this to happen in CI.
2. We want others to have easy time reusing our build configuration -> meaning they configure as little parameters as possible.

So how do we achieve both? I don't have much knowledge on point (2), you do though, so it is best that we follow your advice there -> shared by default.

But to satisfy point (1), we need to build both static and shared library on all OSs, and also test that they can be correctly linked with the executables/used -> that is why I used static in some executables but shared in others.

Sorry, a lot of questions here, but it seems like this is a good opportunity to get all this stuff working as it should be.

Reply:
1. use two build plans for static and shared. Faster, and better coverage
2. I would check for `default_library != static`. If `default_library` is `shared` or `both`, meson will link to the shared library.

Reply:
1. I feel silly asking this, but what do you mean by two build plans? You mean, on CI, first do the whole thing with static lib, and then shared? I can see how it has better coverage, but how would it be faster, if I am doing the same thing twice? Specifically, I am imagining I would add lib_type dimension to the build matrix.

2. Aha, smart meson! But what about c_flags in library() and compile_flags in dependency, and c_extra_flags in pkg-config, how would you conditionally deal with those, especially in case when it is 'both'?

Reply:
Honestly, `both` likely won't matter on Windows, hence I'd just inject the flags on `!= 'static'` and live with the `cl.exe` warnings in Windows. Most Windows users will want to use static libraries.

To your 2. I would inject them as previously done in this PR.

Reply:
@SoapZA I think I am close to getting what you mean, but I don't think I got it all yet. I will enumerate the questions so we get all of them:

1. Ok, so it is likely best if I set default_library to 'shared' on linux and 'static' on windows. But how do I conditionally set it to 'static' if on windows? How do you check in meson if on windows or not?
2. Should I, in CI, run Meson twice -> once with default_library as 'static' and once with 'shared'? So I am sure they are correctly built and linked and all executables work in both situations?
3. Won't it be problem if default_library is set to 'both' and I pass cpp_args for shared library -> what happens with static lib then, from what I understood previously it will not work? You mentioned cl.exe warning on Windows -> that is it, just warnings? What about Linux? Should I maybe forbid somehow using 'both'? Or I could have one shared_library() and one static_library() but then by checking value of default_library I can trigger only those I want -> kind of like manual version of library but I have more control in case that we are building both.

Reply:
Sorry, let me rephrase
1. You can't. Hence I would just either leave it away (i.e. shared by default) or specify `static` as default. Default options cannot be a function of the environment.
2. That's what I meant. Define two different configs in the matrix and test shared and static separately (I use a load of Bamboo as CI at work, and the terminology there is slightly different).
3. Yes it will, and I'm saying to ignore it. `both` is an extremely obscure config, and very un-Windows-y. Building both static and shared libraries concurrently is a feature that people were expecting from libtool, and for a long time Meson didn't support this configuration. Just add the necessary macros on `!= 'static'` and call the `both` case a lost cause (remember, even in `both`, it just issues a linker warning).

There's one final hack that could make all of this less painful: Remember that **`__declspec(dllimport)` is an optimization, it is not strictly necessary**. It only helps in removing the thunk indirection. Given that most hardcore algorithmics happens inside Edlib, I would think that in 99% of cases the thunk indirection is negligible, i.e., you could just omit the machinery to inject `__declspec(dllimport)`. This would massively simplify the setup, **as consuming code isn't affected anymore, only compile-time code**. What do you think @madebr?

Reply:
The following is reasoning why `dllimport/dllexport` is really needed.

> There's one final hack that could make all of this less painful: Remember that **`__declspec(dllimport)` is an optimization, it is not strictly necessary**.

Let's only consider the 2 main compilers for windows: mingw (gcc) and MSVC:
- [mingw's linker's default is `--export-all-symbols`](https://sourceware.org/binutils/docs/ld/WIN32). Which means all symbols that are not declared `static` are visible/callable from outside the dll).
- MSVC's has no choice, there all symbols are hidden. MSVC does not provide a compiler/linker option to make a symbol public. CMake has [WINDOWS_EXPORT_ALL_SYMBOLS](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html), but this will only export all non-static functions/classes. Global variables are not exported.

[Using dllimport/dllexport is the MSVC way to export symbols from a dll](https://docs.microsoft.com/en-us/cpp/cpp/dllexport-dllimport?view=vs-2019), that is [also supported by mingw](http://www.mingw.org/wiki/exporting_a_class_from_a_dll_for_multiple_instantiations).

When exporting e.g. the function `void __declspec(dllexport) function(void)`, then MSVC will create 2 symbols:
1. one inside the dll, called void function(void)`, which other functions inside the  dll can call
2. a `__imp_function(void)` thunk function inside the import library. ([see stackoverflow}(https://stackoverflow.com/questions/3573475/how-does-the-import-library-work-details) for more details.

When building edlib as a static library, you don't want to add `__declspec(dllexport)` to the functions.
Then, when a consumer builds a shared library, using edlib, the symbols of edlib will also be exported.
Which might be unwanted.

In that regard, @SpaceIm's approach in `edlib.h` is very elegant. It even handles symbol visibility on Linux/Macos.

> It only helps in removing the thunk indirection. Given that most hardcore algorithmics happens inside Edlib, I would think that in 99% of cases the thunk indirection is negligible, i.e., you could just omit the machinery to inject `__declspec(dllimport)`. This would massively simplify the setup, **as consuming code isn't affected anymore, only compile-time code**. What do you think @madebr?

You can also completely get rid of dllimport/dllexport, by creating a `.def` file.
See [msdn's documentation](https://docs.microsoft.com/en-us/cpp/build/reference/module-definition-dot-def-files?view=vs-2019).
[An example of such a file can be found in the repo of libtommath](https://github.com/libtom/libtommath/blob/develop/tommath.def)
msdn advises to use `dllexport/dllimport` because using def files makes it much easier to forget to add a symbol. Also, declaration and export visibility of some symbol is done in one place.

**Meson specific**

I believe meson also support this.
I don't know the meson syntax/options that good (because [its documentation](https://mesonbuild.com/Reference-manual.html) is not good).
Is it something like this what you want?
```
edlib_options = ["-DEDLIB_BUILD"]
if default_library == "shared":
    edlib_options.append("-DEDLIB_SHARED")
    # other conditional code
edlib = library("edlib", edlib_sources, options=edlib_options)
...
```
Btw, meson's documentation also [advises to use `library`](https://mesonbuild.com/Reference-manual.html#library), instead of `shared_library/static_library/both_libraries`.

Reply:
@SoapZA thank you that clears it all up!
Ok, I pushed two more commits. meson.build should now follow your recommendations @SoapZA I believe (pls check). I also updated travis and appveyor to build both shared and static library with meson and to test them both. @madebr you might also be interested to see what meson.build looks like right now.

@madebr why would you take out EDLIB_BUILD from the .cpp and then pass it from the outside? Isn't it simpler to just have it in there so we don't have to think/know about it?

Regarding the rest of the discussion here, it is somewhat over my head so I will let you pplz figure out what is the best way to go. To me personally, current solution looks ok and is mostly understandable even for me who doesn't have much experience with this stuff. Simplication is always great though, but only if it doesn't take more than it gives at the end.

Reply:
> @madebr why would you take out EDLIB_BUILD from the .cpp and then pass it from the outside? Isn't it simpler to just have it in there so we don't have to think/know about it?

It's ok to put `#define EDLIB_BUILD` at the top of a `.c` source files, as long as the symbol is defined before including the header in every(!) source of the library.
I'm against putting it at the top of a source file because, when a second source is added to the library, this define must also be the first define. If it forgotten, linking errors will occur (which will hopefully be detected by CI).
Adding the define once and for all in the build scripts (cmake and meson) will avoid this once and for all.

> Regarding the rest of the discussion here, it is somewhat over my head so I will let you pplz figure out what is the best way to go regarding dllimport. To me personally, current solution looks ok and is mostly understandable even for me who doesn't have much experience with this stuff. Simplication is always great though, but only if it doesn't take more than it gives at the end.

I think the solution suggested by @SpaceIm is both elegant and powerful as it supports both Windows dll's and so/dylib visibility on Linux/Macos. You will have much bigger control in defining your api and blocking users to call private functions.